src/alarms/alarm-handling.md

```diff
@@ -0,0 +1,98 @@
+
+# Data flow
+
+Figure below illustrates the data flow from Customer Application broker/thin-edge up to cloud. It shows especially the use of MQTT QOS=1, MQTT Retain and Mosquitto Persistence to achieve reliability alarm state-change transfer.
+
+![Sequence Diagram Update SW-list](images/alarm_dataflow.svg)
+
+**To be decided:** Interface from Mapper to Cloud to be defined. Two options are possible:
+1) **JSONviaMQTT:** To have reliability use QOS=1. Open issue: Not yet completely confirmed that C8Y sends PUBACK when msg was processed, instead just on arrival.
+2) **HTTP REST:** HTTP response indicates if message was processed successfully. Anyway HTTP REST would increase complexity of implementation and increase data traffic.
+
+## Requirements
+|   |
+|---------------|
+| Thin-edge shall transfer most recent state of an alarm to the cloud. |
+| Other local components (as local processes or connected clients) can consume alarm-states from thin-edge by subscribing local broker topics. A local consumer can be made to process full alarm-state history by using "Clean Start Flag=0", a constant "client Id" and "QOS>0". |
+| The same alarm-state message shall be transferred just once to the cloud.<br/>An alarm-state message is treated as different when at least one field (see section 'Information Set per Alarm' above) in the message differs to the message before.<br/>To avoid duplicates some mapper-specific topic shall be used (retained) to store and identify messages that were already transferred to the cloud (e.g. `mapper/c8y/ack/alarms/<alarm>`). |
+
+
+# Public MQTT-based alarm interface
+
+Similar to thin-edge's measurement interface, alarm interface is based on MQTT topics.
+
+## Topic structure and payload
+
+Proposal to have "severity" and "type-string" as topics:
+
+**thin-edge JSON (alarm) format:**
+```
+Topic:   tedge/alarms/<severity-string>/<type-string>
+Payload: {
+  "message":  <message-string>,
+  "status":   <status-string>,  
+  "time":     <time-string>
+}
+```
+
+**Addressing Child-Devices:**
+To address an alarm to a child-device the sub-topic `childs` followed by the child's "device id" has to be used as below:
+```
+Child-Device Topics:   tedge/alarms/childs/<child-device id>/<severity-string>/<type-string>
+```
+
+Benefit to have "severity" and "type-string" as topics:
+Device-Site reactions to alarm could be easier realized.<br/>
+Examples:
+```
+# Listen to all major alarms on all devices
+tedge mqtt sub "tedge/alarms/+/major/+"
+
+# Listen to specific alarm on all devices
+# (e.g. a bunch of temperature sensors that have each an upper limit setpoint with alarm)
+tedge mqtt sub "tedge/alarms/+/+/upper-limit-exceeded"
+
+# And for sure all combinations.
+```
+
+
+Additionally an also supported "raw"-payload format will allow easy adoption for existing "MQTT publishers":
+
+**Raw payload format:**
+```
+Topic:   tedge/alarms/<severity-string>/<type-string>
+Payload: <message-string>
+```

```

Comment:
Am I missing something here? In the JSON payload we have the field `status-string` but here in the raw payload we suddenly loose this field? 
Is this intentional? Or if this is to use the empty/non-empty message to indicate the status why is it needed in the other payload type?

Reply:
See in the section "Raw payload format":
```
   In raw format for missing fields default values are used as below:
   - "status" := "ACTIVE", if payload is [...]
   [...]
```

Reply:
> Is this intentional? Or if this is to use the empty/non-empty message to indicate the status why is it needed in the other payload type?

Yes, indeed. An empty/no-empty message indicates the status. However I like to keep the status as field in the JSON payload format, since that is the way to provide a timestamp with status CLEARED. If one does not need that, the raw payload format can be used.